File: java/src/main/behavioral/state/Context.java
package behavioral.state;

// Context: Maintains an instance of a State subclass
public class Context {
    private State currentState;

    public Context() {
        this.currentState = new ConcreteStateA(); // Default state
    }

    public void setState(State state) {
        this.currentState = state;
    }

    public void request() {
        currentState.handleRequest(this);
    }

    public State getCurrentState() {
        return currentState;
    }
}

// Example Usage:
// public class Main {
//     public static void main(String[] args) {
//         Context context = new Context();
//
//         context.request(); // Switches from A to B
//         context.request(); // Switches from B to A
//         context.request(); // Switches from A to B again
//
//         // Expected Output:
//         // ConcreteStateA: Handling request. Switching to ConcreteStateB.
//         // ConcreteStateB: Handling request. Switching to ConcreteStateA.
//         // ConcreteStateA: Handling request. Switching to ConcreteStateB.
//     }
// }
//
// Common Mistakes to Avoid:
//
// 1. Forgetting to Update the Context's State:
//    - Mistake: Calling handleRequest() without updating the state.
//    - Fix: Ensure state transitions are correctly implemented.
//
// 2. Allowing Context to Depend on Specific State Implementations:
//    - Mistake: Directly referencing `ConcreteStateA` or `ConcreteStateB` in `Context`.
//    - Fix: Use `State` interface for flexibility.
//
// 3. Not Defining a Default State:
//    - Mistake: Context starts with a `null` state, leading to `NullPointerException`.
//    - Fix: Initialize context with a valid default state.
//
// 4. Overcomplicating Simple Conditionals:
//    - Mistake: Using the State pattern when simple `if-else` logic would suffice.
//    - Fix: Use the pattern when behavior truly varies by state.

File: python/src/behavioral/state/context.py
from .state import State


class Context:
    """
    The Context maintains a reference to a State instance that determines
    its behavior.
    """

    def __init__(self, state: State):
        self._state: State = state

    def change_state(self) -> None:
        """
        Ask the current state what the next state should be.
        """
        self._state = self._state.next_state()

    def request(self) -> str:
        """
        Delegate behavior to current state.
        """
        return self._state.handle()


# Example Usage:
#
# from context import Context
# from concrete_state_a import ConcreteStateA
#
# context = Context(ConcreteStateA())
#
# print(context.request())  # "State A: Handling request."
# context.change_state()
#
# print(context.request())  # "State B: Handling request."
# context.change_state()
#
# print(context.request())  # "State A: Handling request."
#
#
# Common Mistakes to Avoid:
#
# 1. Hardcoding transitions inside Context.
#    - Fix: Transitions belong inside State classes.
#
# 2. Using too many states when if/else would be simpler.
#    - Fix: Use State Pattern only when behavior *must* change dynamically.
#
# 3. Exposing state-switching logic outside Context.
#    - Fix: Provide methods like request() and change_state() only.
